File: java/Lowest Common Ancestor of a Binary Search Tree.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return null;
        int i = Integer.compare(root.val, p.val);
        int j = Integer.compare(root.val, q.val);
        if (i < 0 && j < 0) {
            return lowestCommonAncestor(root.right, p, q);
        } else if (i > 0 && j > 0) {
            return lowestCommonAncestor(root.left, p, q);
        } else {
            return root;
        }
    }
}

File: C++/lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == nullptr) {
            return nullptr;
        }
        if(p->val > q->val) {
            return lowestCommonAncestor(root, q, p);
        }
        if(root->val >= p->val && root->val <= q->val) {
            return root;
        } else if(root->val < p->val) {
            return lowestCommonAncestor(root->right, p, q);
        } else if(root->val > q->val) {
            return lowestCommonAncestor(root->left, p, q);
        }
        return nullptr;
    }
};